src/server/saved_objects/service/create_saved_objects_service.js

```diff
@@ -19,49 +19,13 @@
 
 import { SavedObjectsRepository, ScopedSavedObjectsClientProvider } from './lib';
 import { SavedObjectsClient } from './saved_objects_client';
+import { initializeSavedObjectIndices } from '../migrations';
 
-export function createSavedObjectsService(server) {
-  const onBeforeWrite = async () => {
-    const adminCluster = server.plugins.elasticsearch.getCluster('admin');
-
-    try {
-      const index = server.config().get('kibana.index');
-      await adminCluster.callWithInternalUser('indices.putTemplate', {
-        name: `kibana_index_template:${index}`,
-        body: {
-          template: index,
-          settings: {
-            number_of_shards: 1,
-            auto_expand_replicas: '0-1',
-          },
-          mappings: server.getKibanaIndexMappingsDsl(),
-        },
-      });
-    } catch (error) {
-      server.log(['debug', 'savedObjects'], {
-        tmpl:
-          'Attempt to write indexTemplate for SavedObjects index failed: <%= err.message %>',
-        es: {
-          resp: error.body,
-          status: error.status,
-        },
-        err: {
-          message: error.message,
-          stack: error.stack,
-        },
-      });
-
-      // We reject with `es.ServiceUnavailable` because writing an index
-      // template is a very simple operation so if we get an error here
-      // then something must be very broken
-      throw new adminCluster.errors.ServiceUnavailable();
-    }
-  };
-
+export function createSavedObjectsService(kbnServer, server) {
+  const initializeIndicesOnce = createInitFunction(kbnServer);
   const scopedClientProvider = new ScopedSavedObjectsClientProvider({
     index: server.config().get('kibana.index'),
     mappings: server.getKibanaIndexMappingsDsl(),
-    onBeforeWrite,
```

Comment:
We're no longer passing the onBeforeWrite to the `ScopedSavedObjectsClientProvider` constructor so when we set a custom client factory, we'll no longer have access to the `onBeforeWrite`.

Perhaps a sample usage makes this easier to follow: https://github.com/elastic/kibana/blob/security-app-privs/x-pack/plugins/security/index.js#L120

Reply:
I'm bad with words, so I threw up https://github.com/chrisdavies/kibana/pull/1 to illustrate the changes that would make this possible.